ocrd_tesserocr/recognize.py

```diff
@@ -19,62 +24,251 @@ def process(self):
+            # unrecognised_char |
+            # ok_repeated_ch_non_alphanum_wds -?*=
+            # conflict_set_I_l_1 Il1[]
+            # preserve_interword_spaces 0
+            # tessedit_enable_dict_correction 0
+            # tessedit_enable_bigram_correction 1
+            # stopper_smallword_size 2
+            # wordrec_max_join_chunks 4
+            # suspect_space_level 100
+            # suspect_short_words 2
+            # language_model_ngram_on 0
+            # language_model_ngram_order 8
+            # language_model_min_compound_length 3
+            # language_model_penalty_non_freq_dict_word 0.1
+            # language_model_penalty_non_dict_word 0.15
+            # language_model_penalty_punc 0.2
+            # language_model_penalty_case 0.1
+            # language_model_penalty_script 0.5
+            # language_model_penalty_chartype 0.3
+            # language_model_penalty_spacing 0.05
+            # textord_max_noise_size 7
+            # enable_noise_removal 1
+            # classify_bln_numeric_mode 0
+            # lstm_use_matrix 1
+            # user_words_file
+            # user_patterns_file
             for (n, input_file) in enumerate(self.input_files):
                 log.info("INPUT FILE %i / %s", n, input_file)
                 pcgts = from_file(self.workspace.download_file(input_file))
                 # TODO use binarized / gray
                 pil_image = self.workspace.resolve_image_as_pil(pcgts.get_Page().imageFilename)
                 tessapi.SetImage(pil_image)
+                metadata = pcgts.get_Metadata() # ensured by from_file()
+                metadata.add_MetadataItem(
+                    MetadataItemType(type_="processingStep",
+                                     name=OCRD_TOOL['tools']['ocrd-tesserocr-recognize']['steps'][0],
+                                     value='ocrd-tesserocr-recognize',
+                                     Labels=[LabelsType(externalRef="parameters",
+                                                        Label=[LabelType(type_=name,
+                                                                         value=self.parameter[name])
+                                                               for name in self.parameter.keys()])]))
                 # TODO slow
                 #  tessapi.SetPageSegMode(PSM.SINGLE_LINE)
-                log.info("page %s", pcgts)
-                for region in pcgts.get_Page().get_TextRegion():
+                log.info("Recognizing text in page '%s'", pcgts.get_pcGtsId())
+                page = pcgts.get_Page()
+                if maxlevel == 'page':
+                    # not sure what to do here: 
+                    # - We cannot simply do GetUTF8Text(), because there is no TextEquiv on the page level.
+                    # - We could GetComponentImages(RIL.BLOCK) and add a text region for each, then enter region level recognition below. But what if regions are already annotated? How to go about non-text blocks?
```

Comment:
If textequiv_level is meant as maxlevel (which is semantically cleaner), this should be an error for page as it is implemented.

Reply:
True, although arguably "page" is invalid even if `textequiv_level` was not meant as maxlevel but "level above which to start recognition". (Maybe I am missing possible interpretations here.) Perhaps we should take "page" off the list in `ocrd_tool.json`?